### ISSUE TITLE
Increase object limit (weapons not firing bug)

### ISSUE THREAD

**Issue by wm4** [dxx-rebirth/dxx-rebirth#451]
Descent 2 has a maximum object count of 350. There can be at most than this many robots, power ups, weapon fire, etc. at the same time in the level.

This is a bit restrictive. You _can_ hit the limit in some levels, and then confusion results. A player usually notices this by the fact that shooting weapons does nothing (it will reduce ammo, but not create a laser or a rocket).

At least level 12 of the official descent 2 mission, and several user-made custom levels are close to the maximum object count. D2 level 12 is at about 330 - get "involved" with the robots, and the weapon fire easily makes it hit the hard limit. It gets worse if you drop weapons (that's a strategy I liked to use: drop weapons at the start of the level, so I could continue if I got destroyed at a hard spot). "Descent: Die Hard!" has this problem in level 4. "EQ's Sesame Set" has the problem in almost all later levels (the level author wasn't aware of the runtime implications of going close to the limit).

I suggest to increase the object limit as a major way to improve the user experience. It's trivial (incrementing a constant), but requires some effort to keep save games working. I did it myself on my own copy of the source code (there might be later bug fixes, not sure, but it's been working quite reliably for me): https://github.com/wm4/d2x-c/commit/ca6518d1d343a4abaee6cc071a49c3619942043d

In addition of fixing level 12 (could be called an original game bug), I think it's generally a good idea to give level authors more freedom. I've seen that dxx-rebirth raised some other limits occasionally too.

**Comment by ziplantil** [comment]
Looks like a duplicate of #152

**Comment by wm4** [comment]
Looks like it.

Savegames are no problem, it just requires some more code to handle them. I did it before in my repo.

**Comment by vLKp** [comment]
Save games _are_ a problem, if you want old builds to keep loading new saves.  If you don't care about backward compatibility, then yes, just break the savegame format and move on.  Rebirth has been generally very cautious about breaking backward compatibility.  That said, breaking compatibility for this would render #449 irrelevant.

**Comment by wm4** [comment]
I've mentioned multiple time that I made this change without breaking old savegames,
With some more effort you could even make it write old savegames if the highest used object slot is below 350.

**Comment by vLKp** [comment]
The commit you cite unconditionally adds a new counter before `Point_segs`, and bumps the savegame version.  How does that not break old builds loading new save games?  Old builds can only load saves laid out exactly like they expect.  Any new content inserted in places not skipped by the old build seeing the new version number will break things.

**Comment by wm4** [comment]
Yes, that's correct. The way I implemented it, new builds can always load old savegames, but old builds never can load savegames produced by new builds. I didn't consider it worth the effort (what for?).

But it's still possible to have 2-way compatibility. Simply write the old savegame format if all objects are within the old limit. This requires you to pass along the target savegame version to these write functions, and add ifs where necessary (the save path will look like the load path). This will write compatible savegames except in levels that don't work properly anyway.

Of course, old builds will never be able to load savegames with more than 350 objects.

**Comment by vLKp** [comment]
As I read your change, old builds will offer to let users _try_ to load old savegames, but it will either crash or otherwise behave badly, because you added fields in unexpected places.

> what for?

I've received reports where people expect to save in a new build and load in an old build.  I don't see much value in this, but I've tried to support it where I can, including multiple changes to generate savegames that work around loader bugs in older builds.  Mostly, these were cases where a logically 16-bit value, such as an object index, was extended to 32-bits and written into the save, and the old loader expects to treat it as a 32-bit value.  When `object_none` (`0xffff`) is written, the old loader treats it as `0x0000ffff`, then crashes when that isn't a valid index.  Older builds would sign-extend `0xffff` into `0xffffffff`, which the older loader handled properly.  Newer builds would default to zero-extending `0xffff`.

The ticket that @ziplantil cited already discusses the problems with doing this.  It's not just for levels that "don't work properly anyway."  Consider a regular level that works properly now, including ones you cite that are difficult, but not impossible, due to the object limit.  If the user saves while the level is quiet, particularly if many robots have been destroyed, the save will be an old-style save with a low object count.  If the user saves in the middle of a heavy battle, where there are more than old `MAX_OBJECTS` objects in play, then a new-style save is written.  Users will likely find it surprising that the portability of the savegame depends on the circumstances of the save.

If old builds had better error handling and would gracefully fail to load saves that exceeded the 350 object limit, I would be more open to generating such saves.  However, the loader code has always been very fragile and easy to crash with bad data.  I don't want to have a published build that generates saves that crash older builds, especially if it conditionally crashes old builds depending on the state of the level at the time of the save.  If I'm going to break the savegame format, I want to do it in a way that users cannot attempt to load incompatible saves into older builds.

**Comment by wm4** [comment]
Then change the ID (first 4 bytes), which will make old builds error out early. Maybe also add a compatibility check to new builds, so the ID doesn't have to be changed every time.

None of these issues seem unsolvable.

**Comment by wm4** [comment]
Other ideas:
- Show a warning message when saving a game that needs the new format
- Show an error message when more than 350 object slots are used, and deny saving the game

The latter may possibly be annoying for the user, but avoids all issues you mentioned, while still allowing playing levels like this properly.

**Comment by Mako88** [comment]
Another possibility is to go ahead and make the current build's save loading more robust so that any changes to the format in the future don't cause issues, then just make a compatibility break for old builds.

I feel like it's severely limiting to always allow old builds without good save handling to load all future builds saves.

**Comment by vLKp** [comment]
@wm4: yes, I already thought of changing the signature.  If I do that, old builds will report that slot as "Empty" instead of showing an error.  That will be confusing to users, and may lead to users unintentionally overwriting a new-style save because the old-style build claims the slot is free.  If I hadn't repeatedly fielded complaints of old builds choking on new saves, when those new saves are only available from snapshot builds that most distributions aren't even packaging, I wouldn't be so reluctant to break compatibility.  As is, I feel that if the reporting rate is this high for the small number of users who actually run snapshot builds, it will get much worse when distributions pick up a release, _unless_ the break is done in a way that clearly communicates to users that this won't work.

@Mako88: welcome back.  Yes, I agree, this is a bad situation.  A break is coming.  My plan is for it to be a _big_ break, fixing all the things I know are wrong, in one round.  I have some other rearrangements queued ahead of that.  The only future-proofing I can see to do now is to add a check that should have been there all along:
```
    if (save_file_version > version_written_by_this_build) {
        fail();
    }
```
That would allow future extensions by bumping the version number, and prevent fields added in the middle from confusing old clients.  Then this check just needs to be retroactively added to every release that users keep trying to use, so that I can count on it working reliably.

**Comment by zicodxx** [comment]
@vLKp I think that sounds like a good plan. Always thought that a savegame version bump may be needed one day and I felt having as many issues addressed in the process would be the best way to go about it.

And if this is done in a way that makes it clear that old versions cannot handle these savegames (using the version number for example (that would already trigger a message in older versions IIRC)) I think it won't be a problem. Frankly, the issues reported (#449) is more because it wasn't clear that older versions are not supposed to load these saves.

**Comment by vLKp** [comment]
With the caveat that I have only read the code, but not tested it:

As I wrote above, it looks like the savegame loader only checks for _very old_ savegame versions and rejects those.  It has no protection against trying to load a version newer than it understands, which will succeed only if all the new content is at the very end (or is otherwise hidden in regions that older versions treated as dead), and all old content is still where it was in the older version.  That is why I said the version guard needs to be retroactively added to past releases, because currently, raising the version number does _not_ block old builds from loading new saves.

Load and test version:
https://github.com/dxx-rebirth/dxx-rebirth/blob/95abeb0d6428b1abb308281496cf025de98efd3a/similar/main/state.cpp#L1523-L1532

Definition of `STATE_COMPATIBLE_VERSION`:
https://github.com/dxx-rebirth/dxx-rebirth/blob/95abeb0d6428b1abb308281496cf025de98efd3a/similar/main/state.cpp#L91-L98

All later uses of `version` are for deciding whether to expect some field to exist.
